library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.rendering.item.api.client;
+
+import com.mojang.blaze3d.systems.RenderSystem;
+import com.mojang.blaze3d.vertex.BufferBuilder;
+import com.mojang.blaze3d.vertex.BufferRenderer;
+import com.mojang.blaze3d.vertex.VertexFormat;
+import com.mojang.blaze3d.vertex.VertexFormats;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+
+import net.minecraft.client.render.GameRenderer;
+import net.minecraft.client.util.math.MatrixStack;
+
+/**
+ * Contains various helper methods for rendering GUI elements.
+ */
+@Environment(EnvType.CLIENT)
+public final class GuiRendererHelper {
+	private GuiRendererHelper() {
+		throw new RuntimeException("GuiRendererHelper only contains static declarations.");
+	}
+
+	/**
+	 * Renders a solid-colored quad to the screen.
+	 *
+	 * @param matrices the matrices
+	 * @param buffer the buffer builder
+	 * @param x the X position of the quad
+	 * @param y the Y position of the quad
+	 * @param width the width of the quad
+	 * @param height the height of the quad
+	 * @param color the color of the quad
+	 */
+	public static void renderQuad(MatrixStack matrices, BufferBuilder buffer,
+								  int x, int y, int width, int height,
+								  int color) {
+		var mat = matrices.peek().getPosition();
+		int r = (color >> 16) & 0xFF;
+		int g = (color >> 8) & 0xFF;
+		int b = color & 0xFF;
+		int a = (color >> 24) & 0xFF;
+
+		RenderSystem.setShader(GameRenderer::getPositionColorShader);
+		buffer.begin(VertexFormat.DrawMode.QUADS, VertexFormats.POSITION_COLOR);
+		buffer.vertex(mat, x, y, 0.0f).color(r, g, b, a).next();
+		buffer.vertex(mat, x, y + height, 0.0f).color(r, g, b, a).next();
+		buffer.vertex(mat, x + width, y + height, 0.0f).color(r, g, b, a).next();
+		buffer.vertex(mat, x + width, y, 0.0f).color(r, g, b, a).next();
+		BufferRenderer.drawWithShader(buffer.end());
```

Comment:
I can hear jellysquid screaming a thousand screams as someone loops this method and submits 10k draw calls, bringing the FPS down to 0. I would remove the draw call and only set the shader if the current shader is different and draw the buffer as it is. Otherwise this is *extremely* inefficient code and bad for rendering.

Reply:
Yeah, I _really_ need to figure out how to make this not horrible (AKA sharing a single `BufferRenderer`) while also not making it break should someone set a new shader and not set it back.

Reply:
I would look at the `DrawableHelper` class. I do think it covers most of these methods.